src/devdocs2zim/client.py

```diff
@@ -0,0 +1,189 @@
+    version: str = ""
+    # Specific release of the software the documentation is for, if any.
+    release: str = ""
+    # License and attribution information, if any.
+    attribution: str = ""
+
+    @property
+    def slug_without_version(self):
+        return self.slug.split("~")[0]
+
+    def placeholders(self) -> dict[str, str]:
+        """Gets placeholders for filenames."""
+        home_link = ""
+        code_link = ""
+        if self.links is not None:
+            home_link = self.links.home
+            code_link = self.links.code
+
+        full_name = self.name
+        if self.version:
+            full_name += f" {self.version}"
+
+        # properties are inspired by what devdocs uses for their frontend:
+        # https://github.com/freeCodeCamp/devdocs/blob/6caa5eb1b18ab8d34034f319024bd81877035b36/lib/app.rb#L110
+        return {
+            "name": self.name,
+            "full_name": full_name,
+            "slug": self.slug,
+            "version": self.version,
+            "release": self.release,
+            "attribution": self.attribution,
+            "home_link": home_link,
+            "code_link": code_link,
+            "slug_without_version": self.slug_without_version,
+        }
+
+
+class DevdocsIndexEntry(BaseModel):
+    """A link to a document in the sidebar."""
+
+    # Display name for the entry.
+    name: str
+
+    # Path to the entry in the db.json file. This may contain a fragment identifier
+    # linking to an anchor tag e.g. #fragment that would not exist in the db.json file.
+    path: str
+
+    # Name of the type (section) the entry is located under.
+    type: str
+
+    @property
+    def path_without_fragment(self) -> str:
+        """Key in db.json for the file's contents."""
+        return self.path.split("#")[0]
+
+
+class DevdocsIndexType(BaseModel):
+    """A section header for documentation."""
+
+    # Display name for the section.
+    name: str
+
+    # Number of documents in the section.
+    count: int
+
+    # Section slug. This appears to be unused.
+    slug: str
+
+
+class DevdocsIndex(BaseModel):
+    """Represents entries in the /<slug>/index.json file for each resource."""
+
+    # List of entries.
+    entries: list[DevdocsIndexEntry]
+
+    # List of "types" or section headings.
+    # These are displayed mostly in order, except regular expressions are used to sort:
+    # https://github.com/freeCodeCamp/devdocs/blob/e28f81d3218bdbad7eac0540c58c11c7fe1d33d3/assets/javascripts/collections/types.js#L3
+    types: list[DevdocsIndexType]
+
+
+class DevdocsClient:
+    def __init__(self, documents_url: str, frontend_url: str) -> None:
+        self.documents_url = documents_url
+        self.frontend_url = frontend_url
+
+    def read_frontend_file(self, file_path: str) -> str:
+        """Read a file from the devdocs frontend server.
+
+        Parameters:
+          file_path: Path of the file relative to the root.
+        """
+        resp = requests.get(
```

Comment:
SGTM, I wasn't sure about pulling it out because the settings might need to be different between the frontend and documentation server, but I think this approach will work okay.

In the worst case I think we'll end up caching some small files (e.g. CSS) which we might not otherwise but that'll be okay.